### PR TITLE
Make tweedledum a hard requirement

### LINKED ISSUE #6333: Missing tweedledum as required package
### Information

- **Qiskit Terra version**: 0.17.0
- **Python version**: 3.8.8
- **Operating system**: linux

### What is the current behavior?
`qiskit.circuit.classicalfunction.BooleanExpression` fails with message that the `tweedledum` library is required. However, `tweedledum` is not listed a requirement for the `qiskit-terra` package.

### Steps to reproduce the problem
Use API's that instantiate the `BooleanExpression` class. E.g.
```
from qiskit.circuit.library import PhaseOracle
oracle = PhaseOracle('x & ~y')  
```
Leads to error:
```
Traceback (most recent call last):
  File "<ipython-input-97-28069418327a>", line 3, in <module>
    oracle = PhaseOracle('x & ~y')               # previous API: qiskit.aqua.components.oracles.LogicalExpressionOracle
  File "/opt/conda/lib/python3.8/site-packages/qiskit/circuit/library/phase_oracle.py", line 55, in __init__
    expression = BooleanExpression(expression)
  File "/opt/conda/lib/python3.8/site-packages/qiskit/circuit/classicalfunction/boolean_expression.py", line 39, in __init__
    raise MissingOptionalLibraryError(
MissingOptionalLibraryError: "The 'tweedledum' library is required to use 'BooleanExpression compiler'. You can install it with 'pip install tweedledum'."
```

### What is the expected behavior?

No error about missing library should happen.

### Suggested solutions

Add `tweedledum` to `./requirements.txt`.

Related to #5853 and #5830 where `tweedledum` is present in `./requirements-dev.txt`, but not added to `./requirements.txt`

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit switches the tweedledum requirement from being optional to a
hard requirement for installing qiskit-terra. We rely on tweedledum to
synthesize phase oracles which is commonly used functionality and several
issues have been opened. This use of tweedledum will likely continue to
grow so we should just list it as a requirement moving forward.

We originally made it optional because the functionality depending on
tweedledum was isolated to just the classical function compiler which
wasn't widely used. There were also packaging issues in the past where
the available precompiled binaries for tweedledum didn't support all of
our supported environments, but those have been resolved. There is still
an issue for arm64 macOS binaries but Qiskit doesn't have wide support
for that yet (although there is a job publishing universal2 binaries for terra
nothing else supports it yet).

At the same time this commit cleans up the optional requirements list
so that aer is no longer listed there and we add an 'all' extra so that
people can have a simple entypoint to install all the optional extras at
once.

### Details and comments

Fixes #6333
Fixes Qiskit/qiskit#1253

### CODE REVIEW

**Comment by mtreinish** [comment]
I pushed up https://github.com/boschmitt/tweedledum/pull/143 to add wheels for non-x86 architectures, we probably should have those before merging this as we support aarch64 linux. For arm64 macOS last I checked those are still blocked on scikit-build. The fix for that was merged to master: https://github.com/scikit-build/scikit-build/commit/fd17e59c7a8cdf1f9f230900d84f7ff5fe43956f but hasn't been included in a release yet. I think once that is released nothing is stopping tweedledum from publishing universal2 or arm64 macOS wheels too

**Comment by 1ucian0** [comment]
`on hold` until boschmitt/tweedledum#143 gets merged.

**Comment by nonhermitian** [comment]
tweedledum master works on arm64 macs with the master of scikit-build

**Comment by boschmitt** [comment]
> tweedledum master works on arm64 macs with the master of scikit-build

Thanks for the heads up, I will work on automating the CI to build the builds fo the other platforms supported by qiskit.

**Comment by boschmitt** [comment]
Here is the current status: https://github.com/boschmitt/tweedledum/actions/runs/963306146

`tweedledum` will have a hard time supporting `s390x`. The reason lies in its external dependencies, which seems to require a little endian architecture to properly run. I might try to find a workaround, but it seems it won't an easy fix.

**Comment by mtreinish** [comment]
@boschmitt cool, looks good. When you upload non-x86_64 wheels to pypi we can remove the `on hold` from this PR and prepare to merge it.

**Comment by boschmitt** [comment]
I think this won't make it to this release. I found that some tests are not passing on these "weird" architectures. The failing tests are for functionalities that are not use by Qiskit right now, but it would be better to sort things out before releasing. (I will still try to make it, but debugging on the CI or emulated environments takes time)

**Comment by mtreinish** [comment]
@boschmitt is aarch64 working ok? That's the only platform that really matters for this pr (since it's the only architecture terra supports). Ppc64le and s390x are more just nice to have if/when we start fully supporting them in Qiskit.

**Comment by boschmitt** [comment]
> @boschmitt is aarch64 working ok? That's the only platform that really matters for this pr (since it's the only architecture terra supports). Ppc64le and s390x are more just nice to have if/when we start fully supporting them in Qiskit, but aren't blockers for anything.

Today I made it work. The problem was not only big-endianness vs little-endianness, but also some code that seems to not work on any platform other than normal intel/amd64. (The culprit code lies in a dependency of a dependency)

File: requirements.txt
Comment (boschmitt) [inline requirements.txt:14]:
```suggestion
tweedledum>=1.1,<2.0
```

**Review by 1ucian0 — APPROVED** [review]
LGTM! Thanks

**Pull request merged**
